interfaces/kmod/backend.go

```diff
@@ -32,11 +62,104 @@ func (b *Backend) Name() string {
 	return "kmod"
 }
 
+// Setup creates a conf file with list of kernel modules required by given snap,
+// writes it in /etc/modules-load.d/ directory and immediately loads the modules
+// using /sbin/modprobe. The devMode is ignored.
+//
+// If the method fails it should be re-tried (with a sensible strategy) by the caller.
 func (b *Backend) Setup(snapInfo *snap.Info, devMode bool, repo *interfaces.Repository) error {
-	// TODO: get snippets, load modules, create /etc/modules-load.d/snap.modules.conf file
+	snapName := snapInfo.Name()
+	// Get the snippets that apply to this snap
+	snippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name(), interfaces.SecurityKMod)
+	if err != nil {
+		return fmt.Errorf("cannot obtain kmod security snippets for snap %q: %s", snapName, err)
+	}
+
+	// Get the files that this snap should have
+	glob := interfaces.SecurityTagGlob(snapName)
+	content, modules, err := b.combineSnippets(snapInfo, snippets)
+	if err != nil {
+		return fmt.Errorf("cannot obtain expected security files for snap %q: %s", snapName, err)
+	}
+
+	dir := dirs.SnapKModModulesDir
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("cannot create directory for kmod files %q: %s", dir, err)
+	}
+
+	changed, _, err := osutil.EnsureDirState(dirs.SnapKModModulesDir, glob, content)
+	if err != nil {
+		return err
+	}
+
+	if len(changed) > 0 {
+		return loadModules(modules)
+	}
 	return nil
 }
 
+// Remove removes modules config file specific to a given snap.
+//
+// This method should be called after removing a snap.
+//
+// If the method fails it should be re-tried (with a sensible strategy) by the caller.
 func (b *Backend) Remove(snapName string) error {
-	return nil
+	glob := interfaces.SecurityTagGlob(snapName)
+	content := make(map[string]*osutil.FileState)
+	_, _, err := osutil.EnsureDirState(dirs.SnapKModModulesDir, glob, content)
+	return err
+}
+
+// combineSnippets combines security snippets collected from all the interfaces
+// affecting a given snap into a de-duplicated list of kernel modules.
+func (b *Backend) combineSnippets(snapInfo *snap.Info, snippets map[string][][]byte) (content map[string]*osutil.FileState, modules []string, err error) {
+	content = make(map[string]*osutil.FileState)
+
+	// sort apps by name to ensure the order of snippets is deterministic
+	apps := make([]*snap.AppInfo, 0, len(snapInfo.Apps))
+	for _, appInfo := range snapInfo.Apps {
+		apps = append(apps, appInfo)
+	}
+	sort.Sort(byAppName(apps))
+
+	for _, appInfo := range apps {
+		for _, snippet := range snippets[appInfo.SecurityTag()] {
+			// split snippet by newline to get the list of modules
+			for _, line := range bytes.Split(snippet, []byte{'\n'}) {
+				l := bytes.TrimSpace(line)
+				// ignore empty lines and comments
+				if len(l) > 0 && l[0] != '#' {
+					modules = append(modules, string(l))
+				}
+			}
+		}
+	}
+
+	modules = uniqueLines(modules)
```

Comment:
This is a total nitpick but if:
 - you sort the lines
 - don't sort the apps
 - removing duplicates from a sorted array is easier

You should not do this unless you just want to enjoy the coding though :-)